### PR TITLE
[diem-framework][move-stdlib] Use Named Addresses

### PR DESCRIPTION
**Pull request by tnowacki**
## Motivation

- Increase readability of Diem Framework and also clean-up some now un-needed functions

## Test Plan

- rebuilt DF/Stdlib 

## Related PRs

- Built on  #8390, so need to wait for that

### CODE REVIEW

File: language/move-stdlib/modules/addresses.move [hunk 1]
```diff
@@ -0,0 +1,5 @@
+// Declaration of Std address. Needs to be assigned a value depending on the deployment
+
+// TODO the tooling around this tries to actually build the modules... which isn't great
+// This value should be removed
+address Std = 0x1;
```
Comment (tnowacki) [inline language/move-stdlib/modules/addresses.move:5]: @vgao1996 the whole tooling failed since I think the stdlib tries to build itself as a binary? This isn't something we can really support anymore in the named address land since it needs to be configured per deployment. Not sure what to do about this going forward

**Review by tnowacki — COMMENTED** [review]
@vgao1996 a lot of other crazy new files came out when running to rebuild the DF and stdlib, not sure whats up with that

**Comment by vgao1996** [comment]
@tnowacki looking into it right now...

**Comment by vgao1996** [comment]
I tried to run the diem-framework binary with `= 0x1` removed from from both `move-stdlib/modules/addresses.move` and `diem-framework/modules/CoreAddresses.move` and here's what I got:
```

   ┌── /Users/vgao/libra/language/diem-framework/modules/AccountAdministrationScripts.move:2:23 ───
   │
 2 │ module DiemFramework::AccountAdministrationScripts {
   │                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Unable to fully compile and resolve module declaration
   ·
 2 │ module DiemFramework::AccountAdministrationScripts {
   │        -------------  No value specified for address 'DiemFramework'
   │

error: 

   ┌── /Users/vgao/libra/language/diem-framework/modules/AccountCreationScripts.move:1:23 ───
   │
 1 │ module DiemFramework::AccountCreationScripts {
   │                       ^^^^^^^^^^^^^^^^^^^^^^ Unable to fully compile and resolve module declaration
   ·
 1 │ module DiemFramework::AccountCreationScripts {
   │        -------------  No value specified for address 'DiemFramework'
   │

error: 

   ┌── /Users/vgao/libra/language/diem-framework/modules/AccountFreezing.move:2:23 ───
   │
 2 │ module DiemFramework::AccountFreezing {
   │                       ^^^^^^^^^^^^^^^ Unable to fully compile and resolve module declaration
   ·
 2 │ module DiemFramework::AccountFreezing {
   │        -------------  No value specified for address 'DiemFramework'
   │

error: 

   ┌── /Users/vgao/libra/language/diem-framework/modules/AccountLimits.move:3:23 ───
   │
 3 │ module DiemFramework::AccountLimits {
   │                       ^^^^^^^^^^^^^ Unable to fully compile and resolve module declaration
   ·
 3 │ module DiemFramework::AccountLimits {
   │        -------------  No value specified for address 'DiemFramework'
   │

...
```
This is expected right? I mean I would only expect this to work if you fix the generation code so that it binds the addresses when calling the compiler.

I did not observe "a lot of other crazy new files". Note it's normal to see a bunch of files being deleted -- those are the old artifacts and we failed to regerate them.
```
D releases/artifacts/current/modules/000_Signer.mv
 D releases/artifacts/current/modules/001_Errors.mv
 D releases/artifacts/current/modules/002_CoreAddresses.mv
 D releases/artifacts/current/modules/003_DiemTimestamp.mv
 D releases/artifacts/current/modules/004_Roles.mv
 D releases/artifacts/current/modules/005_ValidatorOperatorConfig.mv
 D releases/artifacts/current/modules/006_Vector.mv
 D releases/artifacts/current/modules/007_Option.mv
 D releases/artifacts/current/modules/008_Signature.mv
 D releases/artifacts/current/modules/009_ValidatorConfig.mv
 D releases/artifacts/current/modules/010_SlidingNonce.mv
 D releases/artifacts/current/modules/011_BCS.mv
 D releases/artifacts/current/modules/012_Event.mv
 D releases/artifacts/current/modules/013_DiemConfig.mv
 D releases/artifacts/current/modules/014_DiemSystem.mv
 D releases/artifacts/current/modules/015_Hash.mv
 D releases/artifacts/current/modules/016_FixedPoint32.mv
 D releases/artifacts/current/modules/017_RegisteredCurrencies.mv
 D releases/artifacts/current/modules/018_Diem.mv
 D releases/artifacts/current/modules/019_AccountLimits.mv
...
```

File: language/move-stdlib/modules/addresses.move (hunk 1 above)
Comment (vgao1996) [reply language/move-stdlib/modules/addresses.move:5]: So just to clarify, going forward, how are the addresses supposed to be bound? I mean I know that you can put the assignment in a .move file, but I feel like you might also need to make them specifiable via the movec interface.

A related but somewhat orthogonal issue is the doc generator. I'm not sure if they can handle named addresses. Have you already added named address support to it?

**Comment by vgao1996** [comment]
Okay, so I tried to run the `diem-framework` binary with `= 0x1` removed from only `move-stdlib/modules/addresses.move` and it works. Nothing was updated and I didn't see any "crazy new files" either. I'm not sure I'm replicating the exact issues you ran into, but I can see two potentially problems we'll have to deal with here.
- How the addresses are bound
  - Is it correct that right now, the only way to bind an address to a name is to include an "assignment" (I don't know the right term here... you know what I mean) in a .move file? I'd argue if want your "package" to be deployable to different platforms, such address assignments should not be part of the package itself. 
  - This might explain some of the problems you ran into: you have Std declared in the stdlib, but unbound. This is fine when you build the diem-framework since it sets Std to 0x1. It is problematic however when you build stdlib alone (we currently do this to generate the docs). 
  - I wonder if you should allow the addresses bindings to specified via the public interfaces of movec.
- How do we present addresses in the docs/making the doc generator support named addresses
  - Should we have named or actual addresses in the docs? Also a similar question is, should the docs be platform-specific?

**Comment by tnowacki** [comment]
> How the addresses are bound

Yeah you need an assignment currently, but you can declare it without binding it. I want to have `address Std;` in this file, not `address Std = 0x1;` 
You can declare a address any number of times you want, so over in the DF it has a declaration `address Std = 0x1;` which is fine for these packaging issues. 
I do plan on allowing addresses to be specified as an argument for `move-build`, but I don't really plan on doing that anytime soon. 

> How we present addresses in the docs

Named addresses. Always. 

Addresses are unique to their name, you cannot swap it out for the value in the syntax. So if you have `address Std = 0x42;` you have to do `Std::Vector`, and `0x42::Vector` is an error. 

I don't see why docs would have to be deployment specific based on this. 

But overall, why do the docs need the compiled code? Doesn't make a lot of sense to me.

**Comment by tnowacki** [comment]
> I did not observe "a lot of other crazy new files". Note it's normal to see a bunch of files being deleted -- those are the old artifacts and we failed to regerate them.

Thats probably what I saw, that on one of my commits the files were deleted when they were failed to be regenerated, and then got added again later once it worked. Makes sense, thanks!

File: language/diem-framework/modules/CoreAddresses.move [hunk 2]
```diff
@@ -1,50 +1,21 @@
-address 0x1 {
+address DiemFramework = 0x1;
+address Std = 0x1;
```
Comment (tzakian) [inline language/diem-framework/modules/CoreAddresses.move:2]: Should this be an undeclared value here, and have its value picked up from the move-stdlib?

File: language/diem-framework/modules/doc/CoreAddresses.md [hunk 3]
```diff
@@ -5,17 +5,12 @@
 
 Module providing well-known addresses and related logic.
 
-> Note: this module currently defines zero-argument functions like <code><a href="CoreAddresses.md#0x1_CoreAddresses_DIEM_ROOT_ADDRESS">Self::DIEM_ROOT_ADDRESS</a>()</code> using capitalization
```
Comment (tzakian) [inline language/diem-framework/modules/doc/CoreAddresses.md:8]: Can you possibly add support for named addresses to the documentation generator? Or is this not part of the move-model yet?

**Review by tzakian — COMMENTED** [review]
The changes look good to me. Left a few smaller comments. 

The main question I have is whether we want `Std` or `StdLib` for the move-stdlib address name. I'm leaning towards `StdLib` since that seems more precise, but I don't have a super strong opinion here.

**Comment by bob-wilson** [comment]
I personally prefer "Std" over "StdLib", and along the same lines, I'm wondering if "Diem" would be better than "DiemFramework". In both cases, the "Lib" or "Framework" suffix seems unnecessary and just makes the names longer. To be clear, I'm not suggesting we change the "Diem Framework" name, only that the named address doesn't need to include the "framework" part.

File: language/diem-framework/modules/CoreAddresses.move (hunk 2 above)
Comment (tnowacki) [reply language/diem-framework/modules/CoreAddresses.move:2]: The stdlib shouldn't assign a value for it, the framework should since it is using the stdlib in its deployment. In other words, move-stdlib should never specify it

File: language/diem-framework/modules/doc/CoreAddresses.md (hunk 3 above)
Comment (tnowacki) [reply language/diem-framework/modules/doc/CoreAddresses.md:8]: I'll see what I can do

**Comment by tnowacki** [comment]
> I personally prefer "Std" over "StdLib", and along the same lines, I'm wondering if "Diem" would be better than "DiemFramework". In both cases, the "Lib" or "Framework" suffix seems unnecessary and just makes the names longer. To be clear, I'm not suggesting we change the "Diem Framework" name, only that the named address doesn't need to include the "framework" part.


Definitely agreed on the `Std` case. 

For the framework, my main reason of adding `Framework` all over is there are other `Diem` associated addresses. 

Like it felt weird to have 
```
address Diem = 0x1;
address DiemRoot = 0xA550C18;
```
and 
```
address DiemFramework = 0x1;
address DiemRoot = 0xA550C18;
```
felt a bit better

File: language/diem-framework/modules/CoreAddresses.move (hunk 2 above)
Comment (tzakian) [reply language/diem-framework/modules/CoreAddresses.move:2]: ahhh very good point! I had it turned around in my head.

**Comment by bob-wilson** [comment]
> But I agree it is a bit verbose and makes everything longer... maybe just `DF`?

`DF` is too short for my tastes. `Diem` and `DiemRoot` seems fine to me.

**Comment by tzakian** [comment]
To me `Diem` seems too terse. The point that Todd makes is a good one, and I think having descriptive (and consistent) names is important here. 

One other thing is if at a later date the association wanted to publish modules just for itself, then having the `Diem` name for that would make more sense, and the Diem Framework to be under the `DiemFramework` address.

While I have preference for `DiemFramework`, I'm not strongly against `Diem` so would be fine with that if we can't reach consensus on `DiemFramework` (but am pretty strongly against `DF` -- it's too short and not descriptive enough for me).

**Comment by bob-wilson** [comment]
I can live with `DiemFramework`

**Comment by vgao1996** [comment]
I'd also prefer "DiemFramework" over "DF" -- just imagine having both "DF" and "DiemRoot" and they look inconsistent to me. You could rename the latter to "DR" but I would still prefer us being explicit.

**Review by tzakian — DISMISSED** [review]
LGTM!

**Comment by lerencao** [comment]
sir, when will this land?

**Comment by tnowacki** [comment]
@lerencao, sorry for the delay but this has hit a bit of snag. There is an issue in that the sandbox for the CLI (among other tools) uses compiled modules to build fake "interface files" for the source language compiler. In that translation from source code ~> compiled modules ~> source code, the named addresses are lost and I am working on an extension to the compiler (and the sandbox) to address this. 

So it might be a bit, sorry!

File: language/move-prover/boogie-backend/src/lib.rs [hunk 4]
```diff
@@ -35,8 +35,8 @@ const VECTOR_SMT_ARRAY_EXT_THEORY: &[u8] =
     include_bytes!("prelude/vector-smt-array-ext-theory.bpl");
 const MULTISET_ARRAY_THEORY: &[u8] = include_bytes!("prelude/multiset-array-theory.bpl");
 
-const BCS_MODULE: &str = "0x1::BCS";
-const EVENT_MODULE: &str = "0x1::Event";
+const BCS_MODULE: &str = "Std::BCS";
```
Comment (wrwg) [inline language/move-prover/boogie-backend/src/lib.rs:38]: Undo this change should resolve the Boogie issue.

Comment (tnowacki) [reply language/move-prover/boogie-backend/src/lib.rs:38]: Thanks for the spot! Definitely went a bit overboard with the find/replace

**Review by tzakian — APPROVED** [review]
